### PR TITLE
Add clear Apply and Cancel buttons to Link Control

### LINKED ISSUE #46593: Add clear Submit and Cancel buttons to Link UI
Currently there a number of problems in the Link UI which result from a lack of clarity about when a user intends to "submit" changes to a given link.

A classic example is the "Open in new tab" toggle. Should toggling this close the UI and update the link? Or should it merely toggle the switch and wait for you to re-submit the entire link.

It would be a lot clearer and unambiguous if we simply introduced `Submit` and `Cancel` buttons. Here's a nice example of Evernote:

Screen Shot 2022-12-15 at 21 23 46 (1)

By adding these we can remove an entire class of bugs whilst at the same time making the UX a lot clearer for users.

### PR DESCRIPTION
**Pull request by getdave**
## What?

Adds `Apply` and `Cancel` buttons to the `<LinkControl>` component.

Affords the ability to explicitly cancel the creation or editing of the link.

Closes https://github.com/WordPress/gutenberg/issues/46593

## Why?

**TLDR** 
- it needs to be clearer how to commit and cancel creation/editing of links. Currently it's unclear and unintuitive. 
- Addition of clear unambiguous buttons greatly improves the experience. 
- Specific "Close" button is _required_ for a11y reasons (click away or cancel on blur is [_not_ sufficient](https://github.com/WordPress/gutenberg/pull/46933#issuecomment-1386433792)).

There are a number of small UX issues with not having clear buttons to apply or cancel a link. Specifically there have been quite a few issues reporting around inconsistencies and bugs when submitting changes and UX confusion about how/when changes should be committed.

Examples

- https://github.com/WordPress/gutenberg/issues/45741
- https://github.com/WordPress/gutenberg/issues/43144
- https://github.com/WordPress/gutenberg/issues/40892
- https://github.com/WordPress/gutenberg/issues/35795

This PR is a first step in addressing these issues. See `Next Steps` for followup ideas.

#### Editing an existing link

When you have an existing link and you make changes to the text or URL it's not immediately obvious how to submit those changes. In fact you have to click on the very small icon button which is inside the URL input. Questions that might arrise:

- Where on earth(!) is the submit button? 
- Does clicking submit on the URL input submit my changes to the Text value of the link?

By adding a clear `Apply` button it's 

- clear where the button is that "commits" the changes
- clear that one button will commit _all_ the changes.

#### Cancelling edits to an existing link

When you have an existing link and you make changes to the text or URL and then decide you don't want to commit those changes it's very unclear how to exit the editing process.

In fact you have to click outside of the component to close it. But that's not at all intuitive.

With a dedicated `Cancel` button it is immediately clear how to exit the editing process and all ambiguity is removed.

#### Cancelling the creation of a new link

When you are creating a new link and then decide that in fact you do not wish to do this it's not immediately clear how to exit the link creation process.

In fact you have to click outside of the component to close it. But that's not at all intuitive and can even lead to focus loss from the place where you were originally creating the link which is extremely frustrating.

With a dedicated `Cancel` button it is immediately clear how to exit the editing process and all ambiguity is removed. Moreover your focus is returned back to where you were originally editing which avoids UX issues related to that.

## How?

Adds x2 new buttons 

- `Apply` - this replaces the icon-base submit button but displays all of the same behaviours and allows for submitting any current changes to the link value.
- `Cancel` - this is new and provides a clear means for the user to cancel editing or creation of links.

Care has been taken to ensure that when cancelling an action any _un_committed edits to a link value are discarded.

## Next Steps

If this PR lands then the next step would be to make changes to the link settings (e.g. `Open in new tab`) to be explicitly committed by the user. 

Currently these controls require lots of custom code to handle when they should/shouldn't be committed and it's never resulted in a satisfactory UX. 

We can greatly improve the UX simply by having a rule than _any_ changes to the link value must be explicitly committed bu the user.

#### Manual testing

- New Post
- Add some text
- Create a new link
- See new UI with `Apply` and `Cancel` buttons
- Click `Cancel` - see link creation is cancelled and focus returned to previous location.
- Create a new link and submit a value.
- Click the edit button on the link you created.
- Make some text changes and URL changes then click `Cancel`.
- See the changes you made have been discarded. 

Repeat the above with other locations where `LinkControl` is used such as

- Nav block
- Button block
- Image block

...etc

#### Automated testing

Review and then run the tests:
```
npm run test:unit packages/block-editor/src/components/link-control/test/index.js
```

### Testing Instructions for Keyboard

Repeat same tests as above but ensure that you can complete the tasks using the keyboard only.

Specifically check that the tab order of the Link Control allows for `Apply` as the first tab stop (and thus the primary option) but also allows access to `Cancel` as the secondary stop.

## Screenshots or screencast 

Screenshot 2023-01-06 at 13 35 07

#### Before

https://user-images.githubusercontent.com/444434/210999164-8a24578e-9783-4916-86bc-915d12ec574c.mp4

#### After

https://user-images.githubusercontent.com/444434/211000861-61950caa-803c-443a-bcec-788cd6d5f43e.mp4

### CODE REVIEW

**Comment by getdave** [comment]
If anyone notices the misaligned labels on the control I fixed that in https://github.com/WordPress/gutenberg/pull/46936

**Comment by draganescu** [comment]
Just a quick take: I like the substrate of this PR and I agree with the problem it tries to solve, illustrated well in the 1st video. However I think the two buttons are just too much. The improvements seem to be more likely in:

- reflecting visible updates when they happen
- adding a smaller, secondary way to exit with no change, maybe some [x]

**Comment by getdave** [comment]
FYI I've had [an Issue up for while about this](https://github.com/WordPress/gutenberg/issues/46593) but didn't make much effort to get it traction.

**Comment by getdave** [comment]
> Just a quick take: I like the substrate of this PR and I agree with the problem it tries to solve, illustrated well in the 1st video. However I think the two buttons are just too much. The improvements seem to be more likely in:
> 
> * reflecting visible updates when they happen
> * adding a smaller, secondary way to exit with no change, maybe some [x]

I'm no designer. However whatever route design suggests it needs to cover the scenarios outlined above. Currently everything relies on "discovering" how to do things.

IMHO it should be _clear_ without having to learn or undue cognitive load, even if that's at the expense of aesthetics.

A critical point to note is that (in the future) I'm suggesting making _all changes_ to links require the user to explicitly submit them. The key example is that I should be able to toggle the `Open in new tab` switch and not have that applied until I click `Apply`.

If you look at the code for how all the various implementations of "link creation" consume `<LinkControl>` you'll note that each handles the toggle scenario slightly differently. These minor inconsistencies cause bug reports and everyone expects it to behave differently depending on their own context.

Standardising on "user must explicitly commit all changes to links" makes it a lot clearer. 

> However I think the two buttons are just too much.

Just noting that the x2 buttons only appear when editing. Here are some examples in the wild that use clear buttons:

Screen Shot 2022-12-15 at 21 23 46 (1)

Screen Shot 2023-01-06 at 13 30 07

> adding a smaller, secondary way to exit with no change, maybe some [x]

IMHO a `x` is not equivalent to "Cancel". It means dismiss but it's still ambiguous about what will happen to any changes that have been made.

**Comment by draganescu** [comment]
> A critical point to note is that (in the future) I'm suggesting making all changes to links require the user to explicitly submit them. The key example is that I should be able to toggle the Open in new tab switch and not have that applied until I click Apply.

I think this sort of thing is not expected anymore for quite some time, because UX has more and more discarded apply in favor of easy undo. Apply should be for destructive or hard to undo things, like a sequence of updates where you can lose track of what you did. A toggle is on when you toggled it on in just about all places from system settings of OSes to browser settings and so on.

**Comment by jameskoster** [comment]
I also appreciate the sentiment of this change and agree the link editing experience can feel a bit ambiguous. But I'm also not sure that the solution is as simple as universally adding these two buttons. 

One thing that works well at the moment is that clicking a search result is equivalent to an "Apply". It's a nice confirmatory action and it would be a shame to introduce any additional clicks there.

Taking a step back, the problem seems to be that non-confirmatory closes of the link control do not feel predictable in terms of end result. IE if you click outside the popover there's no indication as to whether your changes were saved or not. This is particularly bad when editing a link because there's no visual feedback on the canvas.

All that considered, an alternative idea to try:

Screenshot 2023-01-06 at 16 17 53

Clicking a result (or pressing enter) is the only way to confirm a selection, doing so closes the link control. The Cancel button will cancel the addition of a new link, or discard any edits to an existing one. Consequently there are only two actions that can close the link control, and both feel predictable. Having typed that out I wonder if a popover is the correct pattern here, and whether a modal would work better.

**Comment by richtabor** [comment]
> I also appreciate the sentiment of this change and agree the link editing experience can feel a bit ambiguous. But I'm also not sure that the solution is as simple as universally adding these two buttons.

Agree. 

> One thing that works well at the moment is that clicking a search result is equivalent to an "Apply". It's a nice confirmatory action and it would be a shame to introduce any additional clicks there.

I would think "Apply" and clicking on a result would both complete the same action. 

> With a dedicated Cancel button it is immediately clear how to exit the editing process and all ambiguity is removed.

It does add clarity, though clicking elsewhere to cancel/close a popover is a standard UX practice. If it's unfinished/unconfirmed, then it resets. 

What about if we add just the Apply button, and if one does not click "Apply", press "Enter" (which is how it works today), or click on one of the search results, then neither an initial link - or link edit - should save?

**Comment by richtabor** [comment]
> Clicking a result (or pressing enter) is the only way to confirm a selection, doing so closes the link control. The Cancel button will cancel the addition of a new link, or discard any edits to an existing one. 

My suggestion above is basically the opposite :)

No "Cancel", but better confirmation via "Apply" (still keep the results, and enter flow).

**Comment by richtabor** [comment]
Secondary, but we should also fix this: 

CleanShot 2023-01-06 at 11 57 23@2x

**Comment by jameskoster** [comment]
> No "Cancel", but better confirmation via "Apply" (still keep the results, and enter flow).

What would the flow to add a link look like? It seems to be:

1. Click add link tool
2. Search / type url
3. Click result

In this case wouldn't the "Apply" button be redundant? 

For this to work I suppose the "Apply" button might only appear when editing an existing link. But still, if clicking a result is akin to clicking "Apply" then it still seems redundant 🤔 

> though clicking elsewhere to cancel/close a popover is a standard UX practice

It is, but I think the problem here is that when you do so it's unclear whether the result is a `cancel` or `save` action? You have to open the link control again to check. It's not so bad when adding a link because you get visual feedback to confirm the link was added (link styles appear). But if you're editing a link there's no such feedback.

---

Another idea: _Adding_ a link works as it does now. _Editing_ a link takes place in a modal. Very crude mockup:

Screenshot 2023-01-06 at 18 09 27

**Comment by getdave** [comment]
> Secondary, but we should also fix this:
> 
> CleanShot 2023-01-06 at 11 57 23@2x

You must have missed https://github.com/WordPress/gutenberg/pull/46933#issuecomment-1373496210 ?

**Comment by getdave** [comment]
This one is quite complex. There are a number of different scenarios to consider other than just creating a new link.

I will say that this PR does retain the existing functionality which is when you select an option it automatically selects that link (no need for `Apply` here).

However when _editing_ a link or creating a link that's a _URL_ (e.g. you type in `www.wordpress.org`) you _need_ a means to submit. That's why we _already_ have a Submit button on `trunk`. 

Therefore I'd say the main changes to consider in this PR are really

- making the `Apply` button much more obvious (as I said it _already_ exists).
- introducing an explicit `Cancel` action.

Therefore the questions we need to consider are:

- is an explicit `Cancel` button better than a potentially ambiguous requirement to click outside to close?
- is a clear `Apply` button better than a small icon button for "submit".

I appreciate this is a difficult problem so I appreciate everyone chipping in here to aid in improving the experience. It's badly needed for the reasons I've stated in the PR desc.

**Comment by scruffian** [comment]
I agree that we need different things when editing vs creating a new link. For example if I'm creating a new link as below, it's hard to imagine a situation when we'd ever need those buttons:
Screenshot 2023-01-06 at 21 21 57

However when editing a link the need is quite clear:
Screenshot 2023-01-06 at 21 22 07

I think this is also wrapped up in some of the changed being discussed at https://github.com/WordPress/gutenberg/pull/46891/files#top, where we also want to consider changing the experience for editing links, so a more thorough design exploration would be beneficial.

**Comment by richtabor** [comment]
> What would the flow to add a link look like? It seems to be:

>Click add link tool
>Search / type url
>Click result
>In this case wouldn't the "Apply" button be redundant?


Yes. I think it's fine. You're still confirming the action, either by clicking a result - or adding a custom url and applying it. 


> It is, but I think the problem here is that when you do so it's unclear whether the result is a cancel or save action? 

I'm thinking it should only be saved if you've taken action in the popover. I.e. You've applied a custom link, or selected a page via the results.

**Comment by alexstine** [comment]
This is one thing that annoys me about Gutenberg and other applications that work in similar ways. I do not want to guess if my action was committed or canceled, having a clearly defined set of buttons for this eliminates all mystery. I was never in favor of putting visuals above good UX. Where do we draw the line... I think this is a good path to go down and hopefully these common sense controls can be brought to other components. I'll A11Y review this when I have a moment.

Thanks.

**Comment by alexstine** [comment]
@getdave Really really enjoy this UX. One thought. Is it possible to rearrange the flow slightly?
1. Input field for URL.
2. Options.
3. Action buttons.

With the action buttons in tab order before options, they could be missed.

Thanks.

**Comment by paaljoachim** [comment]
Actually I like the clarity a cancel and apply button gives. (Even though I have a feeling that something is not entirely there yet.) It gives a much stronger clarity and does not assume the user knows how to handle the link control. 
Instead of pressing esc on the keyboard or clicking outside the link control to cancel, one can just click the cancel button. Instead of pressing the return key or clicking the return icon one can click the apply button. Additional brain storming is needed as it is not fully in place yet. I just am not able to yet place my finger on it.

**Comment by getdave** [comment]
Thanks for all the input. My summary of it is:

- some folks think we should rely on implicit actions such as click outside to cancel as this preserves the visual aesthetics
- others feel strongly that using clear buttons provides a better experience for a wider range of users as things are less ambiguous
- folks have mentioned that buttons when _creating_ a link might be over the top. But I noted that this is probably only the case when you are searching rather than when you are entering a URL manually.
- folks noted that some additional design refinement might be in order here
- a11y changes to content ordering have been suggested.

Next steps:

- get additional input in UX and visual design from @WordPress/gutenberg-design.
- address a11y suggestions

**Comment by getdave** [comment]
I just realised a great argument for clear "Apply" and "Cancel" buttons is for contexts where `<LinkControl>` isn't displayed within a popover. 

In this case we can't rely on "clicking outside". The UI needs to be a lot more explicit.

Here's an example of that in the new Nav offcanvas experiment where I'm using `<LinkControl>` within the inspector controls where there is no concept of a "popover".


https://user-images.githubusercontent.com/444434/211296626-e9e27b82-2eaf-4031-bdc5-0e6b1d75ab48.mp4

**Comment by jameskoster** [comment]
We do have popovers in the Inspector, for example selecting a color:

Screenshot 2023-01-09 at 13 59 03

Interesting to note: Clicking away from the popover there feels like a natural confirmation, and cancelling isn't a big deal, especially since we have the undo tool.

This makes me think that a more robust design for the link control UI might be a better solution here.

**Comment by getdave** [comment]
> Interesting to note: Clicking away from the popover there feels like a natural confirmation, and cancelling isn't a big deal, especially since we have the undo tool.

Just noting that relying on "click away to close" in the inspector tools can lead to you loosing context because the click ends up being in the canvas and you accidentally select another block. I've done this so many times and it's super annoying.

> a more robust design for the link control UI might be a better solution here. 

I'd love to collaborate with someone on this 🙇

**Comment by jasmussen** [comment]
Here's what Google Docs does:

Screenshot 2023-01-13 at 09 36 13

It has an apply button to the right of the link input, with Esc for cancelling. In my personal experience, that feels like it strikes a balance between ergonomics and clarity. How does that fare with you all?

**Comment by getdave** [comment]
@jasmussen Thanks for your feedback. I've noted the Google Docs example (I should have included that in the Issue). We could definitely look to make an incremental improvement to make the "Apply" button clearer like they have done. That said I still have concerns about the lack of a clear "close" action. For example, if this control does not appear within a popover I don't think we can rely on clicking away as a means of cancelling any edits made in the control and using `ESC` is definitely a power user option in my opinion. We can try and iterate on this by first adding a clear "Apply" button and then seeing if that helps things.

@joedolson @alexstine Do you have any a11y feedback on this initiative? Folks seem to prefer not having 2 clear buttons but rather a single clearer "Apply" and leaving the "Cancel" / "Close" action to be _implicit_ via `ESC` or clicking away. As folks seem to be against a visible `Close` button we can include a hidden `Close` button for screen readers if that would help?

@alexstine I note you have said ["ESC" is not a given in feedback on another similar PR](https://github.com/WordPress/gutenberg/issues/47016#issuecomment-1381284570) regarding the popover component.

**Comment by joedolson** [comment]
As somebody who has personally failed to actually add a link using the link control many, many times, I'm 100% in favor of the apply button. That would be awesome.

I do agree that the ability to cancel and retain block selection is important - that's also something that I've been frustrated with on occasion. I would strongly prefer a visible cancel button, so that both available actions are explicit.

A hidden cancel button could be OK, as long as it became visible on focus, but that's something that's going to be inequal for users, since sighted mouse users would not get the benefit. 

There are existing models for cancelling actions that are less obtrustive - the 'x' close button on a modal is generally understood to imply 'cancel', and could also be used here. Overall, cancelling is a less critical action than applying, since it's the default behavior for most actions that aren't 'Apply', and all of those would still be available.

**Comment by alexstine** [comment]
@getdave Here is how I see it.
1. Apply button is required because communicating to users how to commit changes makes the application much more verbose. This makes screen reader UX even worse when a simple apply button gets rid of this issue all together.
2. I would be okay with either a close button and cancel button or one or the other. Escape is not a given and clicking outside is not possible for screen reader users. Just saying the word "click" should automatically make developers/designers understand that this is not an accessible solution to the problem.

I think as long as these buttons are not in the way, designers can simply factor these in to the UI. This feels like a large visual discussion at this point. I cannot be much more help. Bottom line, expecting users to know how to perform an action based on some undiscoverable JavaScript event is NOT good UX for anyone. Visually impaired, blind, fully sighted, or even lower performing computer users.

Thanks.

**Comment by getdave** [comment]
@jasmussen Is this the route you are suggesting? Basically just making Apply much clearer.

Screenshot 2023-01-17 at 02 55 51

We still need to consider the comments above regarding the need for a clear Close button. I'm not sure how to solve that other than the route I've already suggested in this PR. Very open to other ideas.

One idea might be adding the cancel button on the right side opposite the "Open in new tab" toggle.

**Comment by jasmussen** [comment]
> We still need to consider the comments above regarding the need for a clear Close button. I'm not sure how to solve that other than the route I've already suggested in this PR. Very open to other ideas.

For me it seems worth extracting that note and looking at it separately, because the URL dialog is not the only popover. So the question is, do all popovers need an explicit close button? 

Screenshot of a paragraph of text, with the block toolbar shown above, and a the More Options dropdown/popover opened

If that's the case, we should look into a uniform solution that's consistent across all those popovers, so screen reader pathfinding doesn't need to deviate between. And in this case, I wonder: would a close button that becomes visible on-focus, like the skip link at the end of the inspector, work in this case?

**Comment by alexstine** [comment]
@jasmussen Yes, every pop over/dialog/modal should have a close button. I think it is fine for it to appear on focus, but having it always there visually is the best case if it fits in the designs.

**Comment by getdave** [comment]
Thank you for that clarity Alex. Perhaps we could start with a solution for the popover used on Link Control and then roll this out more widely if/when it's working well?

Also (not related to Alex's comment) specifically for this Link Control discussion, we should remember that the Link Control can (and will) exist outside of a popover. Therefore it needs its own independent affordance to commit/cancel and action. If this _also_ triggers a popover to close then that's fine so long as it's given the appropriate content for assistive tech.

**Comment by SaxonF** [comment]
Looks good to me @getdave .

**Comment by richtabor** [comment]
Functionally this works well — there are some design bits to tailor, but I'm ok following up on that. 
 
We're currently using `ButtonGroup`, which is intended to be when the buttons are grouped together visually (border radius only on the outside edges). Outside of that UX pattern, the buttons render inconsistently (funky border radii and border colors). 

Resting: 

CleanShot 2023-01-18 at 15 59 03@2x

Focus states:  

CleanShot 2023-01-18 at 15 54 14@2x

CleanShot 2023-01-18 at 15 54 06@2x

For reference, here's how the ButtonGroup is typically used (Image block): 

CleanShot 2023-01-18 at 16 01 54@2x

**Review by richtabor — APPROVED** [review]
Left a comment detailing some initial follow-up, but functionally this is a solid pass. https://github.com/WordPress/gutenberg/pull/46933#issuecomment-1386637244

**Comment by getdave** [comment]
Looks like it broke the e2e tests. Looking into that now...

Ok it's cos we have a new tab stop for `Cancel`. We'll just need to narrow down where in the tests that is and fix them.

**Comment by getdave** [comment]
Some snapshots also seem incorrect. ` target="_blank"` _should_ be in the markup but isn't in the snapshot.

**Comment by getdave** [comment]
Ok should be fixed.

File: packages/block-editor/src/components/link-control/index.js [hunk 1]
```diff
@@ -235,6 +243,29 @@ function LinkControl( {
...
+
+		// Ensure that any unsubmitted input changes are reset.
+		resetInternalValues();
+
+		if ( hasLinkValue ) {
+			// If there is a link then exist editing mode and show preview.
+			stopEditing();
+		} else {
+			// If there is no link value, then remove the link entirely.
+			onRemove?.();
```
Comment (getdave) [inline packages/block-editor/src/components/link-control/index.js:263]: I am now uncertain as to whether this is a good change. 

We are assuming that "cancel" when there is no link means "remove the link". But is that a decision the _component_ should be making?

Perhaps rather we should simply be calling `onCancel` and allow the _consumer_ to decide how that should be handled. I think it's _likely_ that 99% of the time the passed `onCancel` handler will basically be the same as `onRemove` but should we be _assuming_ that?

Comment (scruffian) [reply packages/block-editor/src/components/link-control/index.js:263]: It feels like a fair assumption to me.

Comment (getdave) [reply packages/block-editor/src/components/link-control/index.js:263]:
```suggestion
			onCancel?.();
```

We could do this but make the default value of `onCancel` be `onRemove`. That way we default to `onRemove` but still allow the consumer to optionally change that if they wish.

File: packages/block-library/src/navigation-link/edit.js [hunk 2]
```diff
@@ -598,7 +579,7 @@ export default function NavigationLinkEdit( {
 							onClose={ () => setIsLinkOpen( false ) }
 							anchor={ popoverAnchor }
 							hasCreateSuggestion={ userCanCreate }
-							onRemove={ removeLink }
+							onRemove={ () => removeBlocks( clientId ) }
```
Comment (getdave) [inline packages/block-library/src/navigation-link/edit.js]: I don't think we should do this in _this_ PR. It has knock on consequences. For example, it will not allow folks to add empty link items which I believe was functionality that was specifically requsted by users in the past.

The use case is that you might want to build out the structure of a Nav without necessarily adding links to each item. 

With this change any nav link block that doesn't add a link using the initial Link UI popup will be removed.

Happy to consider this but I think it should be in a separate PR to avoid over burdening this PR 🙏

**Review by scruffian — APPROVED** [review]
LGTM

File: packages/e2e-tests/specs/editor/various/__snapshots__/links.test.js.snap [hunk 3]
```diff
@@ -68,6 +68,6 @@ exports[`Links should contain a label when it should open in a new tab 1`] = `
 
 exports[`Links should contain a label when it should open in a new tab 2`] = `
 "<!-- wp:paragraph -->
-<p>This is <a href=\\"http://wordpress.org\\">WordPress</a></p>
+<p>This is <a href=\\"http://wordpress.org\\" target=\\"_blank\\" rel=\\"noreferrer noopener\\">WordPress</a></p>
```
Comment (getdave) [inline packages/e2e-tests/specs/editor/various/__snapshots__/links.test.js.snap:71]: I believe this snapshot has been invalid in `trunk` for a while. This is because the test sets open in new tab to it's active state which should result in the attributes being added to the HTML.

File: packages/block-editor/src/components/link-control/index.js (hunk 1 above)
Comment (getdave) [reply packages/block-editor/src/components/link-control/index.js:263]: Let's leave as is.

File: packages/block-editor/src/components/link-control/index.js [hunk 4]
```diff
@@ -235,6 +243,29 @@ function LinkControl( {
 		}
 	};
 
+	const resetInternalValues = () => {
```
Comment (draganescu) [inline packages/block-editor/src/components/link-control/index.js:246]: Nit: do we need this function? Seems to be called once.
